pkg/kubernetes/deployment/appsv1/v1alpha1/rollout_status.go

```diff
@@ -0,0 +1,152 @@
+	PredicateNotSpecSynced: func(d *deploy) string {
+		return "Deployment rollout in-progress : deployment rollout in-progress : waiting for deployment spec update to be observed"
+	},
+}
+
+// rolloutChecks contains a group of predicate it uses predicateName as key.
+var rolloutChecks = map[predicateName]predicate{
+	// PredicateProgressDeadlineExceeded refer to predicate IsProgressDeadlineExceeded.
+	PredicateProgressDeadlineExceeded: IsProgressDeadlineExceeded(),
+	// PredicateOlderReplicaActive refer to predicate IsOlderReplicaActive.
+	PredicateOlderReplicaActive: IsOlderReplicaActive(),
+	// PredicateTerminationInProgress refer to predicate IsTerminationInProgress.
+	PredicateTerminationInProgress: IsTerminationInProgress(),
+	// PredicateUpdateInProgress refer to predicate IsUpdateInProgress.
+	PredicateUpdateInProgress: IsUpdateInProgress(),
+	// PredicateNotSpecSynced refer to predicate IsSyncSpec.
+	PredicateNotSpecSynced: IsNotSyncSpec(),
+}
+
+// rolloutOutput struct contains message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"isRolledout"`
+	Message     string `json:"message"`
+}
+
+// rawFn is a typed function that abstracts
+// conversion of rolloutOutput struct to raw byte
+type rawFn func(r *rolloutOutput) ([]byte, error)
+
+// rawFn is a typed function that abstracts
+// conversion of rolloutOutput struct
+type asRolloutOutputFn func(r *rolloutOutput) (*rolloutOutput, error)
+
+// rollout enables getting various output format of rolloutOutput
+type rollout struct {
+	output          *rolloutOutput
+	raw             rawFn
+	asRolloutOutput asRolloutOutputFn
+}
+
+// rolloutBuildOption defines the
+// abstraction to build a rollout instance
+type rolloutBuildOption func(*rollout)
+
+// NewRollout returns new instance of rollout meant for
+// rolloutOutput. caller can configure it with different
+// rolloutOutputBuildOption
+func NewRollout(opts ...rolloutBuildOption) *rollout {
+	r := &rollout{}
+	for _, o := range opts {
+		o(r)
+	}
+	r.withDefaults()
+	return r
+}
+
+// withOutputObject sets rolloutOutput in rollout instance
+func withOutputObject(o *rolloutOutput) rolloutBuildOption {
+	return func(r *rollout) {
+		r.output = o
+	}
+}
+
+// withDefaults sets the default options of rolloutBuilder instance
+func (r *rollout) withDefaults() {
+	if r.raw == nil {
+		r.raw = func(o *rolloutOutput) ([]byte, error) {
+			return json.Marshal(o)
+		}
+	}
+
+	if r.asRolloutOutput == nil {
+		r.asRolloutOutput = func(o *rolloutOutput) (*rolloutOutput, error) {
+			return o, nil
+		}
+	}
+}
+
+// Raw returns raw bytes outpot of rollout
+func (r *rollout) Raw() ([]byte, error) {
+	if r.output == nil {
+		return nil, fmt.Errorf("Unable to get rollout status output")
+	}
+	return r.raw(r.output)
+}
+
+// AsRolloutOutput returns rolloutOutput struct as output
+func (r *rollout) AsRolloutOutput() (*rolloutOutput, error) {
```

Comment:
We might need to re-design this as per earlier review comments

Reply:
done